### PR TITLE
Fix gen lexer word boundary, case insensitive, and literal matching cases (plus conformance tests)

### PR DESCRIPTION
**Pull request by klondikedragon**
@alecthomas - I wanted to get your feedback early on the approach to add a specialized test to the conformance test suite. It uses a special token (e.g., `WBTEST:`) to push into a new lexer state that then might have different rules vs the rest of the grammar. This would allow for testing different areas of the lexer without having to add more and more rules to the Root state (which might start interacting with each other over time as the number of conformance test cases grow and the rules get more complex).

Opening as a draft to get your feedback on the approach. This demonstrates how the generated lexer currently doesn't support the word boundary `\b` regex properly. (Will next see what might be needed to fix this in the lexer generation.)

This also tweaks `TestLexerConformanceGenerated` so that if you run this test function from VSCode IDE it won't just hang forever (the command that eventually gets run before this change was `go test -run TestLexerConformanceGeneratedInternal -tags generated '-test.run=^TestLexerConformanceGenerated$'`, which confuses go and causes it to hang forever until Ctrl+C is pressed).

### CODE REVIEW

**Comment by klondikedragon** [comment]
Upon further inspection, the previous iteration of the word boundary conformance test was failing because case insensitive matching wasn't implemented. Updated to fix case insensitive matching and add new tests for this case. And then changed the word boundary test to not rely on case insensitive matching.

**Comment by klondikedragon** [comment]
Identified and fixed another edge case where literal matches were not working if they were at exactly the end of the string.

Also identified and fixed the issue with matching `\b` if it could match immediately before the current position `p` instead of right after. This is common in keyword patterns where you might have something like `\b(SELECT|FROM|...)\b` and want to ensure that both before and after the keyword is a word boundary.

File: cmd/participle/gen_lexer_cmd.go [hunk 1]
```diff
@@ -188,10 +188,14 @@ func generateRegexMatch(w io.Writer, lexerName, name, pattern string) error {
 	// Fast-path a single literal.
 	if len(flattened) == 1 && re.Op == syntax.OpLiteral {
 		n := utf8.RuneCountInString(string(re.Rune))
-		if n == 1 {
-			fmt.Fprintf(w, "if p < len(s) && s[p] == %q {\n", re.Rune[0])
+		if re.Flags&syntax.FoldCase != 0 {
+			fmt.Fprintf(w, "if p+%d <= len(s) && strings.EqualFold(s[p:p+%d], %q) {\n", n, n, string(re.Rune))
```
Comment (alecthomas) [inline cmd/participle/gen_lexer_cmd.go:192]: Nice!

File: lexer/internal/conformance/conformance_test.go [hunk 2]
```diff
@@ -18,6 +18,9 @@ var conformanceLexer = lexer.MustStateful(lexer.Rules{
 	"Root": {
 		{"String", `"`, lexer.Push("String")},
 		// {"Heredoc", `<<(\w+)`, lexer.Push("Heredoc")},
+		{"LiteralTest", `LITTEST:`, lexer.Push("LiteralTest")},
```
Comment (alecthomas) [inline lexer/internal/conformance/conformance_test.go:21]: Yeah nice, I like this pattern.

File: lexer/internal/conformance/conformance_test.go [hunk 3]
```diff
@@ -92,6 +113,68 @@ func testLexer(t *testing.T, lex lexer.Definition) {
...
+		{"OneLiteralAtEnd", `LITTEST:ONE`, []token{
+			{"LiteralTest", "LITTEST:"},
+			{"LITOne", "ONE"},
+		}},
+		{"KeywordLiteralAtEnd", `LITTEST:SELECT`, []token{
+			{"LiteralTest", "LITTEST:"},
+			{"LITKeyword", "SELECT"},
+		}},
+		{"LiteralMixed", `LITTEST:hello ONE test LIKE world`, []token{
+			{"LiteralTest", "LITTEST:"},
```
Comment (alecthomas) [inline lexer/internal/conformance/conformance_test.go:147]: Maybe we could make the test harness strip these automatically? ie. the first token iff it ends with `TEST:`

File: lexer/internal/conformance/conformance_test.go [hunk 4]
```diff
@@ -117,7 +200,7 @@ func TestLexerConformanceGenerated(t *testing.T) {
 	args := []string{"test", "-run", "TestLexerConformanceGeneratedInternal", "-tags", "generated"}
 	// Propagate test flags.
 	flag.CommandLine.VisitAll(func(f *flag.Flag) {
-		if f.Value.String() != f.DefValue {
+		if f.Value.String() != f.DefValue && f.Name != "test.run" {
```
Comment (alecthomas) [inline lexer/internal/conformance/conformance_test.go:203]: Nice catch 🤦‍♂️

**Review by alecthomas — APPROVED** [review]
Overall LGTM, just a couple of suggestions.

**Comment by klondikedragon** [comment]
@alecthomas - I did some cleanup, see what you think. Thanks!

**Review by alecthomas — APPROVED** [review]

**Comment by alecthomas** [comment]
Awesome, thanks :)
